File: 3way/problem16/solution.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var equalPairs = function(grid) {
    let rows = {};
    let count = 0;

    for (let i = 0; i < grid.length; i++) {
        let rowStr = JSON.stringify(grid[i]);
        rows[rowStr] = (rows[rowStr] || 0) + 1;
    }

    for (let j = 0; j < grid.length; j++) {
        let column = [];
        for (let i = 0; i < grid.length; i++) {
            column.push(grid[i][j]);
        }

        let columnStr = JSON.stringify(column);
        if (rows[columnStr]) {
            count += rows[columnStr];
        }
    }

    return count;
};

File: 3way/problem16/Solution.java
class Solution {
    public int equalPairs(int[][] grid) {
        Map<String, Integer> rows  = new HashMap<>();
        int count = 0;

        for(int i = 0; i < grid.length; i++){
            String key = Arrays.toString(grid[i]);
            rows.put(key, rows.containsKey(key) ? rows.get(key) + 1 : 1);
        }

        for(int i = 0; i < grid.length; i++){
            int[] column = new int[grid.length];
            for(int j = 0; j < grid.length; j++){
                column[j] = grid[j][i];
            }
            String columnKey = Arrays.toString(column);
            if(rows.containsKey(columnKey)){
                count += rows.get(columnKey);
            }
        }

        return count;

    }
}
